File: src/helpers/benchmark.c
/*****************************************************************
 * Arturo :VM
 * 
 * Programming Language + Compiler
 * (c) 2019-2020 Yanis Zafirópulos (aka Dr.Kameleon)
 *
 * @file: src/helpers/benchmark.c
 *****************************************************************/

#include "arturo.h"

/**************************************
  Methods
 **************************************/

unsigned long long getCurrentTime() {
    struct timeval t;
    struct timezone tzp;
    gettimeofday(&t, &tzp);
    return t.tv_sec*1000000 + t.tv_usec;
}

unsigned long long getCurrentCycles() {
	return __builtin_readcyclecounter();
}

File: src/helpers/benchmark.nim
#=======================================================
# Arturo
# Programming Language + Bytecode VM compiler
# (c) 2019-2024 Yanis Zafirópulos
#
# @file: helpers/benchmark.nim
#=======================================================

#=======================================
# Libraries
#=======================================

import std/monotimes, strutils, times

#=======================================
# Templates
#=======================================
    
template benchmark*(benchmarkName: string, code: untyped) =
    block:
        let t0 = getMonoTime()
        code
        let elapsed = float(ticks(getMonoTime()) - ticks(t0))
        let elapsedStr = (elapsed/1000000.0).formatFloat(format = ffDecimal, precision = 3)
        echo "[benchmark] time: ", elapsedStr, "ms"

template getBenchmark*(code: untyped): float =
    let t0 = getMonoTime()
    code
    (float(ticks(getMonoTime()) - ticks(t0)))/1000000.0